src/placement/pl_map_common.c

```diff
@@ -399,4 +399,94 @@ determine_valid_spares(struct pool_target *spare_tgt, struct daos_obj_md *md,
 	(*current) = (*current)->next;
 }
 
+int
+pl_map_extend(struct pl_obj_layout *layout, d_list_t *extended_list)
+{
+	struct pl_obj_shard	*new_shards;
+	struct failed_shard	*f_shard;
+	d_list_t		*current;
+	uint8_t                 *grp_map;
+	uint32_t                *grp_count;
+	uint32_t                max_fail_grp;
+	uint32_t		new_group_size;
+	uint32_t		grp;
+	uint32_t		grp_idx;
+	int i, j, k = 0;
+	int rc = 0;
+
+	grp_map = NULL;
+	grp_count = NULL;
+
+	/* Empty list, no extension needed */
+	if (extended_list == extended_list->next || layout->ol_grp_size == 1)
+		goto out;
+
+	D_ALLOC_ARRAY(grp_map, (layout->ol_nr / 8) + 1);
+	D_ALLOC_ARRAY(grp_count, layout->ol_grp_nr);
+	if (grp_count == NULL || grp_map == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	i = 0;
+	max_fail_grp = 0;
+
+	current = extended_list->next;
+	while (current != extended_list) {
+		f_shard = d_list_entry(current, struct failed_shard, fs_list);
+		grp = f_shard->fs_shard_idx / layout->ol_grp_size;
+
+		if (isset(grp_map, f_shard->fs_tgt_id) == false) {
+			setbit(grp_map, f_shard->fs_tgt_id);
+			grp_count[grp]++;
+
+			if (max_fail_grp < grp_count[grp])
+				max_fail_grp = grp_count[grp];
+		} else
+			d_list_del_init(&f_shard->fs_list);
+
+		current = current->next;
+	}
+
+
+	new_group_size = layout->ol_grp_size + max_fail_grp;
+	D_ALLOC_ARRAY(new_shards, new_group_size * layout->ol_grp_nr);
+	if (new_shards == NULL)
+		return -DER_NOMEM;
+
+	while (k < layout->ol_nr) {
```

Comment:
easier to understand in this way? " for (i = k = 0; k < layout->ol_nr;)"
if we need to set initial value of loop, I think it's cleaner to just use "for" statement